gcc/rust/parse/rust-parse-impl.h

```diff
@@ -1532,6 +1498,163 @@ Parser<ManagedTokenSource>::parse_macro_rules_def (AST::AttrVec outer_attrs)
     }
 }
 
+// Parses a decl macro 2.0 definition.
+template <typename ManagedTokenSource>
+std::unique_ptr<AST::MacroRulesDefinition>
+Parser<ManagedTokenSource>::parse_decl_macro_def (AST::Visibility vis,
+						  AST::AttrVec outer_attrs)
+{
+  // ensure that first token is identifier saying "macro"
+  const_TokenPtr t = lexer.peek_token ();
+  if (t->get_id () != MACRO)
+    {
+      Error error (t->get_locus (),
+		   "decl macro definition does not start with %<macro%>");
+      add_error (std::move (error));
+
+      // skip after somewhere?
+      return nullptr;
+    }
+  lexer.skip_token ();
+  Location macro_locus = t->get_locus ();
+
+  // parse macro name
+  const_TokenPtr ident_tok = expect_token (IDENTIFIER);
+  if (ident_tok == nullptr)
+    {
+      return nullptr;
+    }
+  Identifier rule_name = ident_tok->get_str ();
+
+  t = lexer.peek_token ();
+  if (t->get_id () == LEFT_PAREN)
+    {
+      // single definiton of macro rule
+      // e.g. `macro foo($e:expr) {}`
+
+      // parse macro matcher
+      Location locus = lexer.peek_token ()->get_locus ();
+      AST::MacroMatcher matcher = parse_macro_matcher ();
+      if (matcher.is_error ())
+	return nullptr;
+
+      // check delimiter of macro matcher
+      if (matcher.get_delim_type () != AST::DelimType::PARENS)
+	{
+	  Error error (locus, "only parenthesis can be used for a macro "
+			      "matcher in decl macro definition");
+	  add_error (std::move (error));
+	  return nullptr;
+	}
+
+      Location transcriber_loc = lexer.peek_token ()->get_locus ();
+      AST::DelimTokenTree delim_tok_tree = parse_delim_token_tree ();
+      AST::MacroTranscriber transcriber (delim_tok_tree, transcriber_loc);
+
+      if (transcriber.get_token_tree ().get_delim_type ()
+	  != AST::DelimType::CURLY)
+	{
+	  Error error (transcriber_loc,
+		       "only parenthesis can be used for a macro transcriber "
+		       "in decl macro definition");
+	  add_error (std::move (error));
+	  return nullptr;
+	}
+
+      AST::MacroRule macro_rule
+	= AST::MacroRule (std::move (matcher), std::move (transcriber), locus);
+      std::vector<AST::MacroRule> macro_rules;
+      macro_rules.push_back (macro_rule);
+
+      return std::unique_ptr<AST::MacroRulesDefinition> (
+	new AST::MacroRulesDefinition (
+	  std::move (rule_name), AST::DelimType::CURLY, std::move (macro_rules),
```

Comment:
```suggestion
	  std::move (rule_name), AST::DelimType::CURLY, {macro_rule},
```

This should allow you to remove the temporary `macro_rules` move, but it's more of a style thing. Feel free to ignore